### PR TITLE
AKCORE-67: Added the support for max delivery count handling in share partition

### PR DESCRIPTION
**Pull request by chirag-wadhwa5**
This PR adds support for max delivery count handling in share partition. When a record is unsuccessfully processed maxDeliveryCount number of times, it will be Archived, instead of made available Available.

Referece: [AKCORE-67](https://confluentinc.atlassian.net/jira/software/c/projects/AKCORE/issues/AKCORE-67?jql=project%20%3D%20%22AKCORE%22%20ORDER%20BY%20created%20DESC)

[AKCORE-67]: https://confluentinc.atlassian.net/browse/AKCORE-67?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: core/src/main/java/kafka/server/SharePartition.java [hunk 1]
```diff
@@ -544,7 +548,7 @@ public CompletableFuture<Optional<Throwable>> acknowledge(String memberId, List<
                             }
                             // Successfully updated the state of the offset.
                             updatedStates.add(updateResult);
-                            if (updateNextFetchOffset) {
+                            if (updateNextFetchOffset && updateResult.state != RecordState.ARCHIVED) {
```
Comment (AndrewJSchofield) [inline core/src/main/java/kafka/server/SharePartition.java:551]: I don't like having this second condition here. Can we only have updateNextFetchOffset set to true in the cases where we want to update the next fetch offset?

File: core/src/main/java/kafka/server/SharePartition.java [hunk 2]
```diff
@@ -654,7 +666,7 @@ public CompletableFuture<Optional<Throwable>> releaseAcquiredRecords(String memb
                             }
                             // Successfully updated the state of the offset.
                             updatedStates.add(updateResult);
-                            localNextFetchOffset = Math.min(offsetState.getKey(), localNextFetchOffset);
+                            if (updateResult.state != RecordState.ARCHIVED) localNextFetchOffset = Math.min(offsetState.getKey(), localNextFetchOffset);
```
Comment (AndrewJSchofield) [inline core/src/main/java/kafka/server/SharePartition.java:669]: The code style in the rest of this method suggests that the if test and the statement should be on separate lines. There's another instance of this too.

File: core/src/main/java/kafka/server/SharePartition.java [hunk 3]
```diff
@@ -973,11 +989,14 @@ private InFlightState tryUpdateState(RecordState newState, boolean incrementDeli
             }
         }
 
-        private InFlightState startStateTransition(RecordState newState, boolean incrementDeliveryCount) {
+        private InFlightState startStateTransition(RecordState newState, boolean incrementDeliveryCount, int maxDeliveryCount) {
             try {
                 rollbackState = new InFlightState(state, deliveryCount);
+                if (newState == RecordState.AVAILABLE && deliveryCount == maxDeliveryCount) {
```
Comment (AndrewJSchofield) [inline core/src/main/java/kafka/server/SharePartition.java:995]: I would use `deliveryCount >= maxDeliveryCount` to be a bit more defensive. If we had a bug in which we failed to archive when the count was equal, we'd at least pick it up the next delivery attempt.

**Review by AndrewJSchofield — CHANGES_REQUESTED** [review]

**Comment by adixitconfluent** [comment]
You should remove some of the TODOs related to implementation of maxDeliveryCount in SharePartition.java. Example - [here](https://github.com/confluentinc/kafka/pull/1162/files#diff-4ef2490e40ca0c63d51bf46caf866690472805e34d3752a919217b28f97045bcR637)

**Review by adixitconfluent — CHANGES_REQUESTED** [review]
You should remove some of the TODOs related to implementation of maxDeliveryCount in SharePartition.java. Example - [here](https://github.com/confluentinc/kafka/pull/1162/files#diff-4ef2490e40ca0c63d51bf46caf866690472805e34d3752a919217b28f97045bcR637)

File: core/src/main/java/kafka/server/SharePartition.java [hunk 4]
```diff
@@ -88,8 +88,8 @@ public RecordState validateTransition(RecordState newState) {
                 throw new IllegalStateException("The state transition is invalid from the current state: " + this);
             }
 
-            if (this == AVAILABLE && newState != ACQUIRED) {
-                throw new IllegalStateException("The state can only be transitioned to ACQUIRED from AVAILABLE");
+            if (this == AVAILABLE && newState == ACKNOWLEDGED) {
```
Comment (adixitconfluent) [inline core/src/main/java/kafka/server/SharePartition.java:91]: Can we have a transition from AVAILABLE to ARCHIVED? I don't think so.

**Review by adixitconfluent — CHANGES_REQUESTED** [review]

File: core/src/main/java/kafka/server/SharePartition.java (hunk 2 above)
Comment (chirag-wadhwa5) [reply core/src/main/java/kafka/server/SharePartition.java]: Missed it, updated the code

File: core/src/main/java/kafka/server/SharePartition.java (hunk 4 above)
Comment (chirag-wadhwa5) [reply core/src/main/java/kafka/server/SharePartition.java]: Thanks for the PR review. My understanding of the existing code was a bit off. You are right here, I have updated the code in the latest commit.

File: core/src/main/java/kafka/server/SharePartition.java [hunk 5]
```diff
@@ -89,7 +89,7 @@ public RecordState validateTransition(RecordState newState) {
             }
 
             if (this == AVAILABLE && newState != ACQUIRED) {
-                throw new IllegalStateException("The state can only be transitioned to ACQUIRED from AVAILABLE");
+                throw new IllegalStateException("The state can only be transitioned to ACQUIRED or ARCHIVED, not ACKNOWLEDGED from AVAILABLE");
```
Comment (adixitconfluent) [inline core/src/main/java/kafka/server/SharePartition.java:92]: We can let go of this change since condition is not changed.

**Review by adixitconfluent — COMMENTED** [review]
LGTM

**Review by AndrewJSchofield — APPROVED** [review]

**Review by adixitconfluent — APPROVED** [review]

**Review by adixitconfluent — APPROVED** [review]

File: core/src/main/java/kafka/server/SharePartition.java [hunk 6]
```diff
@@ -515,6 +515,10 @@ public CompletableFuture<Optional<Throwable>> acknowledge(String memberId, List<
                             inFlightBatch.maybeInitializeOffsetStateUpdate();
                         }
                         for (Map.Entry<Long, InFlightState> offsetState : inFlightBatch.offsetState.entrySet()) {
+
+                            // updateNextFetchOffset needs to be updated for every record iteration,
+                            // as any record in between could have max delivery count limit exceeded
+                            updateNextFetchOffset = batch.acknowledgeType == AcknowledgeType.RELEASE;
```
Comment (apoorvmittal10) [inline core/src/main/java/kafka/server/SharePartition.java:521]: I think we don't require to check again rather skip the update when offset is marked archived.

File: core/src/main/java/kafka/server/SharePartition.java [hunk 7]
```diff
@@ -534,14 +534,23 @@ public CompletableFuture<Optional<Throwable>> acknowledge(String memberId, List<
...
                                         + " for the share partition: {}-{}", offsetState.getKey(),
                                     inFlightBatch, groupId, topicIdPartition);
                                 throwable = new InvalidRecordStateException("Unable to acknowledge records for the batch");
                                 break;
                             }
+                            // If the maxDeliveryCount limit has been exceeded, the record will be transitioned to ARCHIVED state.
+                            // This should not change the nextFetchOffset because the record is not available for acquisition
+                            if (updateResult.state == RecordState.ARCHIVED) {
+                                updateNextFetchOffset = false;
+                            }
```
[the remark this answers is not in the document]
Comment (apoorvmittal10) [reply core/src/main/java/kafka/server/SharePartition.java:575]: Rather updating the `updateNextFetchOffset = false;` to false we can do something below to fix the scenario I mentioned above.

```suggestion
                            // The record/offsets might move to archived state when max delivery count is reached hence skip update for respective offset if required.
                            if (updateNextFetchOffset && updateResult.state != RecordState.ARCHIVED) {
                                localNextFetchOffset = Math.min(offsetState.getKey(), localNextFetchOffset);
                            }
```

**Review by apoorvmittal10 — CHANGES_REQUESTED** [review]

**Review by apoorvmittal10 — APPROVED** [review]
Just see if you can address the suggestion please.

File: core/src/main/java/kafka/server/SharePartition.java (hunk 1 above)
Comment (apoorvmittal10) [reply core/src/main/java/kafka/server/SharePartition.java]: https://github.com/confluentinc/kafka/pull/1162#discussion_r1568909394

**Pull request merged**